sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,286 @@
+            "--log-level" => {
+                let log_level_str = iter.next().ok_or_else(|| error::Error::Usage {
+                    message: "Did not give argument to --log-level".into(),
+                })?;
+                log_level = Some(LevelFilter::from_str(&log_level_str).map_err(|_| {
+                    error::Error::Usage {
+                        message: format!("Invalid log level '{}'", log_level_str),
+                    }
+                })?);
+            }
+
+            _ => {
+                return Err(error::Error::Usage {
+                    message: "unexpected argument".into(),
+                })
+            }
+        }
+    }
+
+    Ok(Args {
+        log_level: log_level.unwrap_or_else(|| LevelFilter::Info),
+    })
+}
+
+fn run() -> Result<()> {
+    let args = parse_args(env::args())?;
+
+    // TerminalMode::Stderr will send all logs to stderr, as sundog only expects the json output of
+    // the setting on stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Stderr)
+        .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    let public_keys = fetch_public_keys_from_imds()?;
+
+    let user_data = UserData::new(public_keys);
+
+    info!("Generating user-data");
+    // Serialize user_data to a JSON string that can be read by the admin container.
+    let user_data_json = serde_json::to_string(&user_data).context(error::SerializeJson)?;
+    debug!("{}", &user_data_json);
+
+    info!("Encoding user-data");
+    // Host container user-data must be base64-encoded.
+    let user_data_base64 = base64::encode(&user_data_json);
+
+    info!("Outputting user-data");
+    // sundog expects JSON-serialized output so that many types can be represented, allowing the
+    // API model to use more accurate types.
+    let output = serde_json::to_string(&user_data_base64).context(error::SerializeJson)?;
+
+    println!("{}", output);
+
+    Ok(())
+}
+
+// Returning a Result from main makes it print a Debug representation of the error, but with Snafu
+// we have nice Display representations of the error, so we wrap "main" (run) and print any error.
+// https://github.com/shepmaster/snafu/issues/110
+fn main() {
+    if let Err(e) = run() {
+        match e {
+            error::Error::Usage { .. } => {
+                eprintln!("{}", e);
+                usage();
+                // sundog matches on the exit codes of the setting generators, so we should return 1
+                // to make sure that this is treated as a failure.
+                process::exit(1);
+            }
+            _ => {
+                eprintln!("{}", e);
+                process::exit(1);
+            }
+        }
+    }
+}
+
+mod error {
+    use snafu::Snafu;
+    fn code(source: &reqwest::Error) -> String {
+        source
+            .status()
+            .as_ref()
+            .map(|i| i.as_str())
+            .unwrap_or("Unknown")
+            .to_string()
```

Comment:
Does this have to be converted to an owned String?  I would have thought snafu would do that for you.  If not, no problem!

Reply:
I'm not positive. This block came from pluto so I'm hesitant to change it. That's certainly an optimization we could look at in a follow-up imds crate.

Reply:
It's the kind of thing that wouldn't compile if it was wrong, so it's safe to test by removing the `to_string()` and changing the return type to `&str`.

Reply:
Oh, this wouldn't work because `request::Error.status()` returns an owned status, not a reference, and there's nowhere for that to live when the function returns, if we're trying to return a reference.  Sorry!